### PR TITLE
Adding Firestore minified int tests

### PR DESCRIPTION
**Pull request by jshcrowthe**
Refactored the integration/firestore suite to test against the minified firestore binary

FIXES ISSUE: #195

### CODE REVIEW

File: integration/firestore/gulpfile.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+function compileIntTests() {
+  const config = require('../../config/webpack.test');
+  return gulp
+    .src(
+      resolve(__dirname, '../../packages/firestore/test/integration/**/*.ts')
+    )
+    .pipe(
+      replace(
+        /import firebase from ('|")[^\1]+firebase_export\1;?/,
```
Comment (mikelehen) [inline integration/firestore/gulpfile.js]: Can you comment this regex?  I am having a real hard time parsing it...

FWIW, the goal with firebase_export.ts was to have a single file we could swap out in the integration tests.  If you're doing this swapping by running a regex over the entire test suite, then I don't think we need firebase_export.ts anymore...  Though, TBH, using firebase_export.ts seems safer to me (if you have a bug in your regex, you might just accidentally run against sources, right)?

If you keep firebase_export.ts can you update the comment in it?  I think it says the file is overwritten by the integration test, which is no longer true...

Comment (jshcrowthe) [reply integration/firestore/gulpfile.js]: Sure I'll add a comment explaining it.

So I'm actually in favor of keeping the `firebase_export.ts` as it gives me a predictable path to replace. If there is a bug in the regex, you are right, it does hit the original source.

I'll update the comment to explain what it is being used for. I'll also refactor the regex a tad to be a little more lenient on whitespace and things.

**Comment by jshcrowthe** [comment]
As far as validation goes, I, manually, placed a `debugger` statement in the initializeApp call of `firebase-app.js` (which is used through your `util/helpers.ts` to set up the firebase instances. Stepped through the tests and watched them reference the minified source as it went through the first 3-4 suites.

And yes, I'll also update the README

File: integration/firestore/gulpfile.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+         * This regex is designed to match the following statement used in our
+         * firestore integratino test suites:
+         * 
+         * import firebase from '../../util/firebase_export';
+         * 
+         * It will handle variations in whitespace, single/double quote
+         * differences, as well as different paths to a valid firebase_export
+         */
+        /import\s+firebase\s+from\s+('|")[^\1]+firebase_export\1;?/,
+        'declare var firebase;'
```
Comment (mikelehen) [inline integration/firestore/gulpfile.js]: I changed this line to 'cowabunga' and the tests still passed.

File: integration/firestore/karma.conf.js [hunk 3]
```diff
@@ -14,57 +14,22 @@
...
+const karmaBase = require('../../config/karma.base');
 
-// Karma configuration
 module.exports = function(config) {
-  config.set({
+  const karmaConfig = Object.assign({}, karmaBase, {
+    // files to load into karma
+    files: [
+      `${path.dirname(require.resolve('firebase'))}/firebase.js`,
+      `${path.dirname(require.resolve('firebase'))}/firebase-firestore.js`,
```
Comment (mikelehen) [inline integration/firestore/karma.conf.js]: I commented out these lines for firebase.js and firebase-firestore.js and the tests still pass.

**Comment by jshcrowthe** [comment]
Huh seems [`webpack-stream`](https://npm.im/webpack-stream) doesn't support streaming so this actually **is** loading the sources... I guess I must have seen a different error.

I'll get this working this morning.

File: integration/firestore/gulpfile.js (hunk 2 above)
Comment (jshcrowthe) [reply integration/firestore/gulpfile.js]: It should be working now.

File: integration/firestore/karma.conf.js (hunk 3 above)
Comment (jshcrowthe) [reply integration/firestore/karma.conf.js]: Should be fixed now.

**Comment by jshcrowthe** [comment]
@mikelehen Ok, for realsies this time. I thing I fixed all the bugs I was also able to catch the error by reverting d49daa0.

File: integration/firestore/gulpfile.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        resolve(__dirname, '../../packages/firestore/**/*.ts'),
+        `!${resolve(__dirname, '../../packages/firestore/node_modules/**/*')}`,
+        `!${resolve(__dirname, '../../packages/firestore/dist/**/*')}`
+      ],
+      { base: '../../packages/firestore' }
+    )
+    .pipe(
+      replace(
+        /**
+      * This regex is designed to match the following statement used in our
```
Comment (mikelehen) [inline integration/firestore/gulpfile.js]: I guess prettier is messing up the formatting on these comments? :-/

File: integration/firestore/gulpfile.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      webpackStream(
+        Object.assign({}, config, {
+          output: {
+            filename: 'test-harness.js'
+          }
+        }),
+        webpack
+      )
+    )
+    .pipe(filter(['**', '!**/*.d.ts']))
```
Comment (mikelehen) [inline integration/firestore/gulpfile.js]: Why is this needed?  I thought the only output is test-harness.js ?

**Comment by mikelehen** [comment]
Big thanks for getting this figured out, by the way. :-)

File: integration/firestore/gulpfile.js (hunk 5 above)
Comment (jshcrowthe) [reply integration/firestore/gulpfile.js]: The transpiling was dropping the .d.ts files from the build into the dist dir which then resulted in duplicated symbols.

File: integration/firestore/gulpfile.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        /**
+      * This regex is designed to match the following statement used in our
+      * firestore integratino test suites:
+      * 
+      * import firebase from '../../util/firebase_export';
+      * 
+      * It will handle variations in whitespace, single/double quote
+      * differences, as well as different paths to a valid firebase_export
+      */
+        /import\s+firebase\s+from\s+('|")[^\1]+firebase_export\1;?/,
```
[the remark this answers is not in the document]
Comment (jshcrowthe) [reply integration/firestore/gulpfile.js]: Consider the following four statements:

1. **Normal Import**
    ```javascript
    import firebase from '../../util/firebase_export';
    ```

1. **Odd Whitespacing**
    ```javascript
    import  firebase from '../../util/firebase_export';
    ```

1. **Differing Path**
    ```javascript
    import firebase from './firebase_export';
    ```
1. **`@firebase/app`**
    ```javascript
    import firebase from '@firebase/app';
    ```

All but the last will properly match because of the specific requirement of the `firebase_export`. Just matching `import firebase` and expanding it to the end of the line is too broad.

File: integration/firestore/gulpfile.js (hunk 4 above)
Comment (jshcrowthe) [reply integration/firestore/gulpfile.js]: Sad day. I'll go fix it.

**Comment by jshcrowthe** [comment]
My pleasure 😄, it's already proving its worth haha
